core/templates/dev/head/pages/exploration_player/AudioPreloaderService.js

```diff
@@ -20,117 +20,112 @@ oppia.factory('AudioPreloaderService', [
+    };
 
-            $scope.chooseDifferentLanguage = function() {
-              $modalInstance.close({
-                shouldDownloadAllAudioInExploration: false,
-                shouldOpenSettingsModal: true
-              });
-            };
-          }]
-      }).result.then(function(result) {
-        if (result.shouldOpenSettingsModal) {
-          // If the user elected to choose a different language, open
-          // the settings modal (later can isolate to a language-only
-          // modal), and on the callback re-open the bandwidth confirmation
-          // modal if the file for the new language hasn't been loaded.
-          AudioTranslationManagerService
-            .showAudioTranslationSettingsModal(function(newLanguageCode) {
-              if (LanguageUtilService.isAutogeneratedAudioLanguage(
-                  newLanguageCode)) {
-                confirmationCallback(newLanguageCode);
-              } else {
-                var newAudioTranslation =
-                  audioTranslationsForContent[newLanguageCode];
-                if (newAudioTranslation && !AssetsBackendApiService.isCached(
-                  newAudioTranslation.filename)) {
-                  _showBandwidthConfirmationModal(
-                    audioTranslationsForContent, newLanguageCode,
-                    confirmationCallback)
-                }
-              }
-            });
-        } else {
-          confirmationCallback(languageCode);
-          if (result.shouldDownloadAllAudioInExploration) {
-            _preloadAllAudioFiles(languageCode);
-          }
-        }
+    var _cancelPreloading = function() {
+      _isRunning = false;
+      AssetsBackendApiService.abortAllCurrentDownloads();
+      _filesBeingDownloaded.forEach(function(file) {
+        _filesToBeDownloaded.push(file);
       });
+      _filesBeingDownloaded = [];
     };
 
     return {
-      init: function() {
-        _init();
+      init: function(exploration) {
+        _init(exploration);
+      },
+      kickOffAudioPreloader: function(
+          sourceStateName, finishedLoadingAudioCallback) {
+        _kickOffAudioPreloader(sourceStateName, finishedLoadingAudioCallback);
+      },
+      isLoadingAudioFile: function(filename) {
+        return _filesBeingDownloaded.indexOf(filename) !== -1;
       },
-      hasPreloadedLanguage: function(languageCode) {
-        return _preloadedLanguageCodes.indexOf(languageCode) !== -1;
+      restartAudioPreloader: function(
+          sourceStateName, finishedLoadingAudioCallback) {
+        _cancelPreloading();
+        _kickOffAudioPreloader(sourceStateName, finishedLoadingAudioCallback);
       },
-      showBandwidthConfirmationModal: function(
-          audioTranslationsForContent, languageCode,
-          confirmationCallback) {
-        _showBandwidthConfirmationModal(
-          audioTranslationsForContent, languageCode,
-          confirmationCallback);
+      isRunning: function() {
+        return _isRunning;
```

Comment:
Does _isRunning ever get set to false when all audio has finished downloading? I can't find where this occurs.

Reply:
Actually the only point of this was the make sure that I don't repeatedly call kickOffAudioPreloader from AudioControlsDirective; it should only be called at the outset of an exploration. So really I might want to change it to something like "hasStarted". Ideally, I can figure out a way to get rid of this altogether though.

Reply:
Oh, I see. Yeah, I think the kickoff can't happen from a directive. The directive should only be able to make audio requests in the middle of a lesson.

Reply:
So, the directive still indirectly calls kickOffAudioPreloader so that it can start BFS'ing from the current state if it hasn't begun downloading the audio file for that state. Is this okay?